lips/lip-4.md

```diff
@@ -0,0 +1,222 @@
+
+As an example, assume a recipient account provides custodial wallets. To send funds to this wallet within Libra:
+
+* The recipient crafts a QR code containing a URI that references the account hosting the custodial wallet, or the host, and a reference to their wallet.
+* The consumer scans the QR code and constructs a transaction and includes within it the subaddress.
+* Once the transaction has been submitted and committed to the blockchain, the host notifies the recipient of the transaction.
+
+---
+# Specification
+---
+
+# The Lifetime of a Transaction Containing Metadata
+
+The first step to submitting a transaction is producing the metadata. The sender first produces a *Libra Canonically Serialized (LCS)* metadata_wrapper consisting of an LCS-serialized MetadataType object:
+
+```
+enum MetadataType {
+  Undefined,
+  GeneralMetadataType(GeneralMetadata),
+  TravelRuleMetadataType(TravelRuleMetadata),
+  UnstructuredStringMetadataType(UnstructuredStringMetadata),
+}
+
+// Used for versioning of general metadata
+enum GeneralMetadata {
+    GeneralMetadataVersion0(GeneralMetadatav0),
+}
+
+struct GeneralMetadatav0 {
+    // Subaddress to which the funds are being sent
+    Option<Vec<u8>> to_subaddress,
+    
+    // Subaddress from which the funds are being sent
+    Option<Vec<u8>> from_subaddress,
+    
+    // Event sequence number of referenced payment
+    Option<u64> referenced_event,
+}
+
+// Used for versioning of travel rule metadata
+enum TravelRuleMetadata {
+    TravelRuleMetadataVersion0(TravelRuleMetadatav0),
+}
+
+struct TravelRuleMetadatav0 {
+    // Off-chain reference_id.  Used when off-chain APIs are used.
+    // Specifies the off-chain reference ID that was agreed upon in off-chain APIs.
+    Option<String> off_chain_reference_id,
+}
+
+struct UnstructuredStringMetadata {
+    // Unstructured string metadata
+    Option<String> metadata,
+}
+```
+
+
+Using the initial example described in the motivation, the merchant whose wallet, 0x5678, is hosted by 0x1234 may post a URI for shoes that cost 20 microlibra.  The purchaser would then submit a transaction containing the following metadata:
+
+```
+0x01, 0x00, 01, 0x5678, 00, 00, 00
+/* general_metadata_type, general_metadata_v0, 
+      to_subaddress_present, to_subaddress,  
+      from_subaddress_not_present, 
+      referenced_event_not_present */
+```
+
+## Submitting a transaction
+
+With the metadata in hand, the sender can now submit a transaction to the Libra block chain via a deposit call in Move:
+
+```
+deposit<LBR>(
+    recipient_address, 
+    amount, 
+    metadata, 
+    metadata_signature);
+```
+
+In the case of the example above, 0x1234 would be the recipient and 20 LBR would be the amount.  Note that metadata_signature must only be present for travel-rule cases between VASPs.
```

Comment:
- Ah, now I see what 0x1234 is now. We might want to use different syntactic conventions for Libra addresses and subaddresses to avoid confusion (e.g., don't use 0x for the subaddress).
- Might want to say: `metadata` and `metadata_signature` are optional for transactions not subject to the travel rule (unless it sets a bad precedent to say the `metadata` is optional)

Reply:
Yeah, the signature is only for TR, but the metadata object itself isn't necessarily related to travel rule always.  You could put a note in there if you wanted.  You could put a refund ID in there.  Let me try to make it clearer